guide/components/provide-inject.md (#150)
vuejs-translations/docs-fr: src/guide/components/provide-inject.md

# Provide / Inject {#provide-inject}

> Cette page suppose que vous avez déjà lu les [bases à propos des composants](/guide/essentials/component-basics). Lisez-les d'abord si vous débutez avec les composants.

## Passage de props en profondeur (Prop Drilling) {#prop-drilling}

> Habituellement, lorsque nous devons transmettre des données du parent à un composant enfant, nous utilisons des [props](/guide/components/props). Cependant, imaginez le cas où nous avons un arbre de composants important, et qu'un composant profondément imbriqué aurait besoin d'acceder à des informations d'un de ses composants parents. Avec seulement des props, nous devrions passer la même prop sur toute la chaîne composants parents :

Reviewer: ```suggestion
Habituellement, lorsque nous devons transmettre des données du parent à un composant enfant, nous utilisons des [props](/guide/components/props). Cependant, imaginez le cas où nous avons un arbre de composants important, et qu'un composant profondément imbriqué aurait besoin d'accéder à des informations d'un de ses composants parents. Avec seulement des props, nous devrions passer la même prop sur toute la chaîne composants parents :
```
As-tu installé l'extension VS Code pour éviter ce genre de faute d'orthographe?

Author: Non pour ce projet de traduction, je commit depuis github directement.

![prop drilling diagram](./images/prop-drilling.png)

<!-- https://www.figma.com/file/yNDTtReM2xVgjcGVRzChss/prop-drilling -->

> Notez que bien que le composant `<Footer>` n'utilise pas du tout de ces props, il doit malgré tout les déclarer et les transmettre uniquement afin que `<DeepChild>` puisse y accéder. S'il y a une chaîne de composants parents plus longue, encore plus de composants seront affectés par le problème. C'est ce qu'on appelle le "props drilling" et ce n'est certainement pas amusant à gérer.

Reviewer: ```suggestion
Notez que bien que le composant `<Footer>` n'utilise pas du tout ces props, il doit malgré tout les déclarer et les transmettre uniquement afin que `<DeepChild>` puisse y accéder. S'il y a une chaîne de composants parents plus longue, encore plus de composants seront affectés par le problème. C'est ce qu'on appelle le "props drilling" et ce n'est certainement pas amusant à gérer.
```

Nous pouvons résoudre le "props drilling" avec `provide` and `inject`. Un composant parent peut servir de **fournisseur de dépendances** pour tous ses descendants. Tout composant enfant de l'arborescence, quelle que soit sa profondeur, peut **injecter** des dépendances fournies par des composants présent dans sa chaîne de composants parents.

![Provide/inject scheme](./images/provide-inject.png)

<!-- https://www.figma.com/file/PbTJ9oXis5KUawEOWdy2cE/provide-inject -->

## Provide {#provide}

<div class="composition-api">

Pour fournir des données aux descendants d'un composant, utilisez la fonction [`provide()`](/api/composition-api-dependency-injection.html#provide) :

```vue
<script setup>
import { provide } from 'vue'

provide(/* clé */ 'message', /* valeur */ 'hello!')
</script>
```

Si vous n'utilisez pas `<script setup>`, assurez-vous que `provide()` est appelé de manière synchrone dans `setup()` :

```js
import { provide } from 'vue'

export default {
  setup() {
    provide(/* clé */ 'message', /* valeur */ 'hello!')
  }
}
```

> La fonction `provide()` accepte deux arguments. Le premier argument est appelé la **clé d'injection**, qui peut être une chaîne de caractères ou un `Symbol`. La clé d'injection est utilisée par les composants descendants pour rechercher la valeur souhaitée à injecter. Un seul composant peut appeler `provide()` plusieurs fois avec différentes clés d'injection pour fournir différentes valeurs.

Reviewer: ```suggestion
La fonction `provide()` accepte deux arguments. Le premier argument est appelé la **clé d'injection**, qui peut être une chaîne de caractères ou un `Symbol`. La clé d'injection est utilisée par les composants descendants pour rechercher la valeur souhaitée à injecter. Un composant peut appeler `provide()` plusieurs fois avec différentes clés d'injection pour fournir différentes valeurs.
```
_A single component_ ne veut pas dire un seul composant, mais ça permet d'appuyer le fait qu'un composant puisse le faire plusieurs fois.

Author: Je suis d'accord, typo here

provide('key', count)
```

> La fourniture de valeurs réactives permet aux composants descendants utilisant la valeur fournie d'établir une connexion réactive au composant fournisseur.

Reviewer: ```suggestion
Fournir des valeurs réactives permet aux composants descendants qui les utilisent d'établir une connexion réactive au composant fournisseur.
```
Plus simple à lire

Author: en effet

}
```

> Pour chaque propriété de l'objet `provide`, la clé est utilisée par les composants enfants pour localiser la valeur correcte à injecter, tandis que la valeur est ce qui finit par être injecté.

Reviewer: ```suggestion
Pour chaque propriété de l'objet `provide`, la clé est utilisée par les composants enfants pour localiser la valeur correcte à injecter, tandis que la valeur est ce qui finit par être injectée.
```
Non?

Author: hmm, non ça s'accorde avec "ce", sinon ce devrait être "la valeur est celle qui finit par être injectée." mais on ne dit jamais ça. Argument supplémentaire ?

Reviewer: Ben pour le coup je suis pour utiliser celle, car on parle bien de la valeur

    }
  },
  provide() {
    // utiliser la syntaxe avec une fonction afin de pouvoir accéder à `this`
    return {
      message: this.message
    }
  }
}
```

Cependant, notez que cela **ne rend pas** l'injection réactive. Nous discuterons ci-dessous de la [manière de rendre les injections réactives](#working-with-reactivity).

</div>

> ## App-level Provide {#app-level-provide}

Reviewer: Hum..
```suggestion
## Provide au niveau de l'application {#app-level-provide}
```
?

</script>
```

Si la valeur fournie est une référence, elle sera injectée telle quelle et **ne sera pas** automatiquement exposée. Cela permet au composant injecteur de conserver la connexion de réactivité au composant fournisseur.

> [Exemple complet provide + inject avec réactivité](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiwgcHJvdmlkZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBDaGlsZCBmcm9tICcuL0NoaWxkLnZ1ZSdcblxuLy8gYnkgcHJvdmlkaW5nIGEgcmVmLCB0aGUgR3JhbmRDaGlsZFxuLy8gY2FuIHJlYWN0IHRvIGNoYW5nZXMgaGFwcGVuaW5nIGhlcmUuXG5jb25zdCBtZXNzYWdlID0gcmVmKCdoZWxsbycpXG5wcm92aWRlKCdtZXNzYWdlJywgbWVzc2FnZSlcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxpbnB1dCB2LW1vZGVsPVwibWVzc2FnZVwiPlxuICA8Q2hpbGQgLz5cbjwvdGVtcGxhdGU+IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9zZmMudnVlanMub3JnL3Z1ZS5ydW50aW1lLmVzbS1icm93c2VyLmpzXCJcbiAgfVxufSIsIkNoaWxkLnZ1ZSI6IjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgR3JhbmRDaGlsZCBmcm9tICcuL0dyYW5kQ2hpbGQudnVlJ1xuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEdyYW5kQ2hpbGQgLz5cbjwvdGVtcGxhdGU+IiwiR3JhbmRDaGlsZC52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IHsgaW5qZWN0IH0gZnJvbSAndnVlJ1xuXG5jb25zdCBtZXNzYWdlID0gaW5qZWN0KCdtZXNzYWdlJylcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxwPlxuICAgIE1lc3NhZ2UgdG8gZ3JhbmQgY2hpbGQ6IHt7IG1lc3NhZ2UgfX1cbiAgPC9wPlxuPC90ZW1wbGF0ZT4ifQ==)

Reviewer: Traduire les commentaires dans le code?

Author: ? Je ne vois pas où ?

Author: Modifer le playground ?? (je ne suis pas pour)

<div class="options-api">

Pour injecter des données fournies par un composant parent, utilisez l'option [`inject`](/api/options-composition.html#inject) :

```js

> export default {

Reviewer: injected value => valeur injectée

}
```

Les injections sont résolues **avant** l'état même du composant, vous pouvez donc accéder aux propriétés injectées dans `data()` :

```js
export default {
  inject: ['message'],
  data() {
    return {
      // données initiales basées sur la valeur injectée
      fullMessage: this.message
    }
  }
}
```

> [Exemple complet provide + inject](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdD5cbmltcG9ydCBDaGlsZCBmcm9tICcuL0NoaWxkLnZ1ZSdcblxuZXhwb3J0IGRlZmF1bHQge1xuICBjb21wb25lbnRzOiB7IENoaWxkIH0sXG4gIHByb3ZpZGUoKSB7XG4gICAgcmV0dXJuIHtcbiAgICAgIG1lc3NhZ2U6ICdoZWxsbydcbiAgICB9XG4gIH1cbn1cbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxDaGlsZCAvPlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59IiwiQ2hpbGQudnVlIjoiPHNjcmlwdD5cbmltcG9ydCBHcmFuZENoaWxkIGZyb20gJy4vR3JhbmRDaGlsZC52dWUnXG5cbmV4cG9ydCBkZWZhdWx0IHtcbiAgY29tcG9uZW50czoge1xuICAgIEdyYW5kQ2hpbGRcbiAgfVxufVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEdyYW5kQ2hpbGQgLz5cbjwvdGVtcGxhdGU+IiwiR3JhbmRDaGlsZC52dWUiOiI8c2NyaXB0PlxuZXhwb3J0IGRlZmF1bHQge1xuICBpbmplY3Q6IFsnbWVzc2FnZSddXG59XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8cD5cbiAgICBNZXNzYWdlIHRvIGdyYW5kIGNoaWxkOiB7eyBtZXNzYWdlIH19XG4gIDwvcD5cbjwvdGVtcGxhdGU+In0=)

Reviewer: Traduire le texte dans `GrandChild`

Author: On ne modifie pas les playgrounds ??

Reviewer: Si, on modifie mais uniquement traduire les commentaires.

> ### Aliasing d'injection {#injection-aliasing}

Reviewer: ```suggestion
### Alias d'injection {#injection-aliasing}
```
Pas sûr de moi mais me paraît plus commun que Aliasing, QDT*?

**Qu'en Dit-Tu?*

> Lors de l'utilisation de la syntaxe utilisant un tableau pour `inject`, les propriétés injectées sont exposées sur l'instance de composant à l'aide de la même clé. Dans l'exemple ci-dessus, la propriété a été fournie sous la clé `"message"` et injectée en tant que `this.message`. La clé locale est la même que la clé d'injection.

Reviewer: ```suggestion
Si la syntaxe utilisée avec `inject` est sous la forme d'un tableau, les propriétés injectées sont exposées sur l'instance de composant à l'aide de la même clé. Dans l'exemple ci-dessus, la propriété a été fournie sous la clé `"message"` et injectée en tant que `this.message`. La clé locale est la même que la clé d'injection.
```

> Si nous voulons injecter la propriété en utilisant une clé locale différente, nous devons utiliser la syntaxe utilisant un objet avec l'option `inject` :

Reviewer: ```suggestion
Si nous voulons injecter la propriété en utilisant une clé locale différente, nous devrions utiliser la syntaxe sous la forme d'un objet avec l'option `inject` :
```

```js
export default {
  inject: {
    /* clé locale */ localMessage: {
      from: /* clé d'injection */ 'message'
    }
  }
}
```

Ici, le composant localisera une propriété fournie avec la clé `"message"`, puis l'exposera en tant que `this.localMessage`.

</div>

### Valeurs par défaut lors de l'injection {#injection-default-values}

Par défaut, `inject` suppose que la clé injectée est fournie quelque part dans la chaîne de composants parents. Dans le cas où la clé n'est pas fournie, il y aura un message d'avertissement lors de l'exécution.

Si nous voulons faire fonctionner une propriété injectée avec des fournisseurs facultatifs, nous devons déclarer une valeur par défaut, de la même manière qu'avec les props :

<div class="composition-api">

```js
// `value` sera "default value"
// si aucune donnée correspondant à la clé "message" n'a été fournie

> const value = inject('message', 'default value')

Reviewer: je me demande si les chaînes de caractères pourraient pas non plus être traduite.. ici `valeur par défaut` ferait du sens

Author: je suis pas pour modifier les exemples au-delà des commentaires. (cela  peut impacter la logique du code, selon comment la chaine est pas la suite manipulée)

</div>

> ## Travailler avec réactivité {#working-with-reactivity}

Reviewer: ```suggestion
## Injections réactives {#working-with-reactivity}
```
ça illustre bien cette partie je pense

<div class="composition-api">

> Lors de l'utilisation de valeurs de réactives avec provide / inject, **il est recommandé de conserver toutes les mutations des l'état réactif à l'intérieur du _fournisseur_ dans la mesure du possible**. Cela garantit que l'état fourni et ses éventuelles mutations sont colocalisés dans le même composant, ce qui facilite sa maintenabilité à l'avenir.

Reviewer: ```suggestion
Lors de l'utilisation de valeurs de réactives avec provide / inject, **il est recommandé de conserver toutes les mutations d'un état réactif à l'intérieur du _fournisseur_ dans la mesure du possible**. Cela garantit que l'état fourni et ses éventuelles mutations sont localisés dans le même composant, ce qui facilite sa maintenabilité à l'avenir.
```
co-localisés me semble un peu répétitif avec dans le même composant.

Il peut arriver que nous ayons besoin de mettre à jour les données depuis un composant réalisant l'injection. Dans de tels cas, nous vous recommandons de fournir une fonction responsable de la mutation de l'état :

```vue{7-9,13}
<!-- à l'intérieur du composant fournisseur -->
<script setup>
import { provide, ref } from 'vue'

> 

Reviewer: Je trouve dommage de ne pas traduire les textes.. Pôle Nord ça le ferait plus nan?

Author: hmm, je ne pense pas, si l'exemple manipule les chaines, le résultat peut être différent si on modifie le contenu. Donc pas fan de cette approche

Reviewer: ➡️ https://github.com/vuejs-translations/docs-fr/discussions/160

  }
}
```

[Exemple complet provide + inject avec réactivité](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdD5cbmltcG9ydCBDaGlsZCBmcm9tICcuL0NoaWxkLnZ1ZSdcbmltcG9ydCB7IGNvbXB1dGVkIH0gZnJvbSAndnVlJ1xuXG5leHBvcnQgZGVmYXVsdCB7XG4gIGNvbXBvbmVudHM6IHsgQ2hpbGQgfSxcbiAgZGF0YSgpIHtcbiAgICByZXR1cm4ge1xuICAgICAgbWVzc2FnZTogJ2hlbGxvJ1xuICAgIH1cbiAgfSxcbiAgcHJvdmlkZSgpIHtcbiAgICByZXR1cm4ge1xuICAgICAgbWVzc2FnZTogY29tcHV0ZWQoKCkgPT4gdGhpcy5tZXNzYWdlKVxuICAgIH1cbiAgfVxufVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtZXNzYWdlXCI+XG4gIDxDaGlsZCAvPlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59IiwiQ2hpbGQudnVlIjoiPHNjcmlwdD5cbmltcG9ydCBHcmFuZENoaWxkIGZyb20gJy4vR3JhbmRDaGlsZC52dWUnXG5cbmV4cG9ydCBkZWZhdWx0IHtcbiAgY29tcG9uZW50czoge1xuICAgIEdyYW5kQ2hpbGRcbiAgfVxufVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEdyYW5kQ2hpbGQgLz5cbjwvdGVtcGxhdGU+IiwiR3JhbmRDaGlsZC52dWUiOiI8c2NyaXB0PlxuZXhwb3J0IGRlZmF1bHQge1xuICBpbmplY3Q6IFsnbWVzc2FnZSddXG59XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8cD5cbiAgICBNZXNzYWdlIHRvIGdyYW5kIGNoaWxkOiB7eyBtZXNzYWdlIH19XG4gIDwvcD5cbjwvdGVtcGxhdGU+In0=)

> La fonction `computed()` est généralement utilisée dans les composants utilisant l'API Composition, mais peut également être utilisée pour compléter certains cas d'usage avec l'API Options. Vous pouvez en savoir plus sur son utilisation en lisant [les principes de base de la réactivité](/guide/essentials/reactivity-fundamentals.html) et [les propriétés calculées](/guide/essentials/computed.html) avec la préférence d'API définie sur API Composition.

Reviewer: ```suggestion
La fonction `computed()` est généralement utilisée dans les composants utilisant la Composition API, mais peut également être utilisée pour compléter certains cas d'usage avec l'Options API. Vous pouvez en savoir plus sur son utilisation en lisant [les principes de base de la réactivité](/guide/essentials/reactivity-fundamentals.html) et [les propriétés calculées](/guide/essentials/computed.html) avec la préférence d'API définie sur Composaition API.
```
Voir Glossaire

:::warning Configuration temporaire requise
L'utilisation ci-dessus nécessite de définir `app.config.unwrapInjectedRef = true` pour que les injections exposent automatiquement les valeurs des références calculées. Cela deviendra le comportement par défaut dans Vue 3.3 et cette configuration est introduite temporairement pour éviter les erreurs de compatibilité. Il ne sera plus nécessaire après 3.3.
:::

</div>

> ## Travailler avec les Symbols en tant que clés {#working-with-symbol-keys}

Reviewer: ```suggestion
## Injection avec des Symbols en tant que clés {#working-with-symbol-keys}
```

> Jusqu'à présent, nous avons utilisé des clés d'injection qui étaient des chaînes de caractères dans les exemples. Si vous travaillez dans une application de taille importante avec de nombreux fournisseurs de dépendances, ou si vous créez des composants qui seront utilisés par d'autres développeurs, il est préférable d'utiliser des clés d'injection utilisant les Symbols pour éviter les collisions potentielles.

Reviewer: ```suggestion
Jusqu'à présent, nous avons utilisé dans les exemples des clés d'injection qui étaient des chaînes de caractères. Si vous travaillez dans une application de taille importante avec de nombreux fournisseurs de dépendances, ou si vous créez des composants qui seront utilisés par d'autres développeurs, il est préférable d'utiliser des clés d'injection utilisant des Symbols pour éviter les collisions potentielles.
```

<div class="composition-api">

```js
// dans le composant fournisseur
import { provide } from 'vue'
import { myInjectionKey } from './keys.js'

> 

Reviewer: _data to provide_ est à traduire

Author: Done

```

```js

> // dans le composant injecteur

Reviewer: ```suggestion
// dans le composant fournisseur
```

Author: hmm, non le composant fournisseur est celui qui utilise `provide` ou alors je vois plutôt changer "composant injecteur" par "composant réalisant l'injection" ?

Reviewer: ha oui pardon.
